tensorflow/lite/kernels/internal/reference/pooling3d.h

```diff
@@ -0,0 +1,193 @@
+  TFLITE_DCHECK_EQ(output_shape.DimensionsCount(), 5);
+
+  ActivationT activation_min, activation_max;
+  GetActivationParams(params, &activation_min, &activation_max);
+  const int batches = MatchingDim(input_shape, 0, output_shape, 0);
+  const int channels = MatchingDim(input_shape, 4, output_shape, 4);
+
+  const int in_spatial_dim_1 = input_shape.Dims(1);
+  const int in_spatial_dim_2 = input_shape.Dims(2);
+  const int in_spatial_dim_3 = input_shape.Dims(3);
+  const int out_spatial_dim_1 = output_shape.Dims(1);
+  const int out_spatial_dim_2 = output_shape.Dims(2);
+  const int out_spatial_dim_3 = output_shape.Dims(3);
+
+  const int stride_spatial_dim_1 = params.stride_depth;
+  const int stride_spatial_dim_2 = params.stride_height;
+  const int stride_spatial_dim_3 = params.stride_width;
+  const int filter_spatial_dim_1 = params.filter_depth;
+  const int filter_spatial_dim_2 = params.filter_height;
+  const int filter_spatial_dim_3 = params.filter_width;
+  const int padding_spatial_dim_1 = params.padding_values.depth;
+  const int padding_spatial_dim_2 = params.padding_values.height;
+  const int padding_spatial_dim_3 = params.padding_values.width;
+
+  for (int batch = 0; batch < batches; ++batch) {
+    for (int out_d1 = 0; out_d1 < out_spatial_dim_1; ++out_d1) {
+      const int in_d1_origin =
+          (out_d1 * stride_spatial_dim_1) - padding_spatial_dim_1;
+      const int filter_d1_start = std::max(0, -in_d1_origin);
+      const int filter_d1_end =
+          std::min(filter_spatial_dim_1, in_spatial_dim_1 - in_d1_origin);
+      for (int out_d2 = 0; out_d2 < out_spatial_dim_2; ++out_d2) {
+        const int in_d2_origin =
+            (out_d2 * stride_spatial_dim_2) - padding_spatial_dim_2;
+        const int filter_d2_start = std::max(0, -in_d2_origin);
+        const int filter_d2_end =
+            std::min(filter_spatial_dim_2, in_spatial_dim_2 - in_d2_origin);
+        for (int out_d3 = 0; out_d3 < out_spatial_dim_3; ++out_d3) {
+          const int in_d3_origin =
+              (out_d3 * stride_spatial_dim_3) - padding_spatial_dim_3;
+          const int filter_d3_start = std::max(0, -in_d3_origin);
+          const int filter_d3_end =
+              std::min(filter_spatial_dim_3, in_spatial_dim_3 - in_d3_origin);
+          for (int channel = 0; channel < channels; ++channel) {
+            ActivationT total = 0;
+            for (int filter_d1 = filter_d1_start; filter_d1 < filter_d1_end;
+                 ++filter_d1) {
+              const int in_d1 = in_d1_origin + filter_d1;
+              for (int filter_d2 = filter_d2_start; filter_d2 < filter_d2_end;
+                   ++filter_d2) {
+                const int in_d2 = in_d2_origin + filter_d2;
+                for (int filter_d3 = filter_d3_start; filter_d3 < filter_d3_end;
+                     ++filter_d3) {
+                  const int in_d3 = in_d3_origin + filter_d3;
+                  total += input_data[Offset(input_shape, batch, in_d1, in_d2,
```

Comment:
Can you replace Offset with the partial calculation of the input index in each of the nested loops? Offset has terrible performance and I am working on getting rid of it.

Reply:
I moved the implementations from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/kernels/pooling3d.cc to their own `pooling3d.h` file. 

As the PR has been open since more than one year and it's an important op to have for us, can we do it in two steps to avoid complexifying this PR more than it's currently? Once merged, I can create an optimisation PR to remove the `Offset` usage if it's alright to you.

Thanks

Reply:
OK for me. Thanks